sdks/go/pkg/beam/core/util/hooks/hooks.go

```diff
@@ -87,14 +87,14 @@ func RunInitHooks(ctx context.Context) error {
 }
 
 // RequestHook is called when handling a FnAPI instruction.
-type RequestHook func(context.Context, *fnpb.InstructionRequest) error
+type RequestHook func(context.Context, *fnpb.InstructionRequest) (context.Context, error)
 
 // RunRequestHooks runs the hooks that handle a FnAPI request.
 func RunRequestHooks(ctx context.Context, req *fnpb.InstructionRequest) {
 	// The request hooks should not modify the request.
 	for n, h := range activeHooks {
 		if h.Req != nil {
-			if err := h.Req(ctx, req); err != nil {
+			if ctx, err := h.Req(ctx, req); err != nil {
```

Comment:
<!--new_thread; commit:de84b439b782a6ad6a18ee9ba1ce4a22852fa63f; resolved:0-->
Should the same context be threaded through each hook, and then the final context returned from RunRequestHooks? At present the new contexts are all thrown away.

Reply:
Fixed.

Reply:
This function seems to still return the original context due to "ctx" defined here shadowing, not overwriting, the argument. Maybe add a test?

Reply:
Yep. Added test. Also added context propagation to init so those hooks can establish the base context for the harness.